File: dotnet/GHIElectronics.DUELink/Analog.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GHIElectronics.DUELink {
    public partial class DUELinkController {

        public class AnalogController {

            SerialInterface serialPort;

            public AnalogController(SerialInterface serialPort) => this.serialPort = serialPort;

            public double Read(int pin) {
                if (pin < 0 || pin >= this.serialPort.DeviceConfig.MaxPinAnalog)
                    throw new ArgumentOutOfRangeException("Invalid pin.");


                var cmd = string.Format("vread({0})", pin.ToString());

                this.serialPort.WriteCommand(cmd);

                var response = this.serialPort.ReadResponse();

                if (response.success) {                   
                    try {
                        var value = double.Parse(response.response);

                        return value;
                    }
                    catch {

                    }


                }

                return 0;
            }

            public bool Write(int pin, double dc) {
                if (pin < 0 || (pin >= this.serialPort.DeviceConfig.MaxPinIO ) ) 
                    throw new ArgumentOutOfRangeException("Invalid pin.");

                if (dc < 0 || dc > 1) {
                    throw new Exception("Dutycle must be in 0..0.1");
                }

                var cmd = string.Format("pwrite({0},{1})", pin.ToString(), dc.ToString());


                this.serialPort.WriteCommand(cmd);

                var res = this.serialPort.ReadResponse();

                return res.success;
            }

            public double ReadVCC() {

                this.serialPort.WriteCommand("ReadVCC()");

                var response = this.serialPort.ReadResponse();

                if (response.success) {
                    try {
                        var value = double.Parse(response.response);

                        return value;
                    }
                    catch {

                    }


                }

                return 0;
            }

        }
    }
}

File: micropython/standard-library/duelink/analog.py

class AnalogController:        
    def __init__(self, serialPort):
        self.serialPort = serialPort

    def VoltRead(self, pin):
        cmd = "vread({0})".format(pin)

        self.serialPort.WriteCommand(cmd)

        r,s = self.serialPort.ReadResponse()

        if r:
            try:
                return float(s)
            except:
                pass

        return 0
    
    def Read(self, pin):
        cmd = "aread({0})".format(pin)

        self.serialPort.WriteCommand(cmd)

        r,s = self.serialPort.ReadResponse()

        if r:
            try:
                return float(s)
            except:
                pass

        return 0
    
    def Write(self, pin, duty_cycle):
               
        if duty_cycle < 0 or duty_cycle > 1:
            raise ValueError('Duty cycle must be in the range 0..1')

        cmd = f'awrite({pin}, {duty_cycle})'
        self.serialPort.WriteCommand(cmd)

        r,s = self.serialPort.ReadResponse()

        return r
    
    def ReadVCC(self):
        cmd = f"readvcc()"
        self.serialPort.WriteCommand(cmd)
        r,s = self.serialPort.ReadResponse()

        if r:
            try:
                return float(s)
            except:
                pass

        return 0
    
